### PR TITLE
[platforms] absorb worker cls difference into platforms folder

### PR DESCRIPTION
**Pull request by youkaichao**
part of https://github.com/vllm-project/vllm/issues/9268 

every platforms should specify the worker class inside their own code.

in addition, the default case is `auto` , and this allows users to specify custom classes for extensibility (which i'm working on as part of an RLHF support).

### CODE REVIEW

**Review by comaniac — APPROVED** [review]
LGTM

**Comment by njhill** [comment]
Thanks @youkaichao! I am also reviewing this now...

**Review by njhill — COMMENTED** [review]
@youkaichao I think this change is probably ok if we're going to continue iterating on the architecture, but it doesn't seem like the right end design to me.

I've already been thinking that we need to overhaul the executor hierarchy/abstractions a bit (hoping we can do this as part of v1), and that may be part of why this doesn't sit right.

In particular we have one or more executor class(es) per platform so this is in some way abstracting over the platforms. But then there's a parallel Platform abstraction. I think we should get rid of the platform-specific executors. I.e. no `ray_*pu_executor.py`s. Possibly the platform-specific aspects could be a mix-in.

It also feels a bit wrong to me to update the config objects in-place since these might be created/ "owned" by the user.

Also wdyt about changing this field to be `custom_worker_cls: Optional[str] = None`? Since it's a very specialized option I would consider it more overriding vLLMs native behaviour so it's not so much an "auto" thing.

**Comment by youkaichao** [comment]
> I think we should get rid of the platform-specific executors. I.e. no ray_*pu_executor.pys. Possibly the platform-specific aspects could be a mix-in.

I strongly agree. We should only have {single worker executor, ray executor, mp executor} , and they should be able to initialize various workers.

I think this PR is a tiny step towards that direction.

File: vllm/platforms/cuda.py [hunk 1]
```diff
@@ -16,6 +16,11 @@
 
 from .interface import DeviceCapability, Platform, PlatformEnum
 
+if TYPE_CHECKING:
+    from vllm.config import VllmConfig
+else:
+    VllmConfig = None
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply vllm/platforms/cuda.py:22]: yes

**Comment by youkaichao** [comment]
> It also feels a bit wrong to me to update the config objects in-place since these might be created/ "owned" by the user.

what's the concern here? the problem is we don't have a scratch space for the model to store some information, and right now we use `vllm_config` a lot, to store per-model basis global variables.

**Comment by njhill** [comment]
> > It also feels a bit wrong to me to update the config objects in-place since these might be created/ "owned" by the user.
> 
> what's the concern here? the problem is we don't have a scratch space for the model to store some information, and right now we use `vllm_config` a lot, to store per-model basis global variables.

Yeah it's kind of a more general point than this PR ... like you say two things are being conflated a bit. Ideally the config should be treated as read-only I think (could have been passed in by the user) and model global mutable state should be separate.

Perhaps something like:
```python
@dataclass
class ModelState:
    config: VllmConfig
    
    # ...
```

**Comment by MengqingCao** [comment]
> Ideally the config should be treated as read-only I think (could have been passed in by the user) and model global mutable state should be separate.

Agree. BTW, I think mutable state like `worker_cls` should only determined by the platform, and maybe we can set it in `xxxPlatform` directly without the `auto` option. Treating it as a configuration item may mislead users to modify it.

**Comment by youkaichao** [comment]
> wdyt about changing this field to be custom_worker_cls: Optional[str] = None?

I use a single `worker_cls` because it will be printed out during init.

Printing:

```text
worker_cls = "vllm.worker.worker.Worker"
custom_worker_cls = "whatever.user.provide"
```

looks less clear than:

```text
worker_cls = "auto"
```

and 

```text
worker_cls = "whatever.user.provide"
```

**Review by njhill — APPROVED** [review]
We can address the other suggestions as follow-on refactoring

**Pull request merged**